### PR TITLE
feat: added shared registries namespace property in model registry dsc component, fixes RHOAIENG-12335

### PR DESCRIPTION
**Pull request by dhirajsb**
## Description

Added string property named registriesNamespace in model registry component DSC config to configue shared registries namespace.
The property can only be set once using CEL validation rules added using kubebuilder annotations in the CRD.
Fixes RHOAIENG-12335

https://issues.redhat.com/browse/RHOAIENG-12335

## How Has This Been Tested?

Includes existing integration test that now takes the registry namespace from the DSC.

## Merge criteria

- [X] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [X] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [X] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by dhirajsb** [comment]
@zdtsw, @VaishnaviHire, @lburgazzoli prow is timing out on pod scheduling, can you kick the failing jobs if you have the rights?

**Comment by dhirajsb** [comment]
Tested the changes in my cluster locally. Verified the following:
* default value `registriesNamespace: odh-model-registries` was automatically added to an existing DSC
* The property `registriesNamespace` is immutable. When I tried editing that property, I got the following error:
```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
# datascienceclusters.datasciencecluster.opendatahub.io "default-dsc" was not valid:
# * spec.components.modelregistry.registriesNamespace: Invalid value: "string": RegistriesNamespace is immutable
#
```

**Review by zdtsw — APPROVED** [review]

**Comment by zdtsw** [comment]
i am fine with the change, just put a "on-hold" to get more eyes on

**Comment by lburgazzoli** [comment]
It would be nice to have a test for the CRD validation so we won't accidentally remove it, otherwise LGTM

**Comment by zdtsw** [comment]
for the e2e, if we can get https://github.com/opendatahub-io/opendatahub-operator/pull/1216 in first, it will be less timeout in further runs

**Review by etirelli — APPROVED** [review]
LGTM, my only question is: where is it setting the default name for RHOAI?

**Comment by zdtsw** [comment]
> LGTM, my only question is: where is it setting the default name for RHOAI?

that will be done in rhoai code and internal build system

**Comment by dhirajsb** [comment]
Also verified in my local cluster that the default value shows up when creating DSC:
image

**Comment by dhirajsb** [comment]
The e2e test is failing due to missing `registriesNamespace` property in `setupDSCInstance()`. I pushed a fix for it.

**Comment by dhirajsb** [comment]
>we should try to avoid this and have the code all in the odh repository, even for rhoai, to avoid those merge issues we discussed before.
The stream specific constant strings like application namespace and model registry namespace are in code strings as well as in docs, and test yaml. 
imho, it's going be a bit tricky to try and pull everything out into a single repo. We might have to use build scripts that surgically replace strings, etc.

**Comment by lburgazzoli** [comment]
My 2c for a future improvement:
- have the model registry controllers to determine the registries namespace from a command line flag and env var, so that the odh operator can inject the right value (but we can keep `odh-model-registries` as default flag value).
- leave the immutability validation on the CRD but implement defaulting in a stream specific webhook (there are many other things that are stream specific other than the model registry namespace that are scattered around code/resources)

**Comment by dhirajsb** [comment]
>have the model registry controllers to determine the registries namespace from a command line flag and env var, so that the odh operator can inject the right value (but we can keep odh-model-registries as default flag value).

We don't need this since the namespace is not passed/used in the model registry controller, only in the odh operator's `modelregistry.go`. 

>leave the immutability validation on the CRD but implement defaulting in a stream specific webhook (there are many other things that are stream specific other than the model registry namespace that are scattered around code/resources)

We can have defaulting webhook set the default values But then we can't have the default value in the CRD description if it's not in the yaml.

**Comment by lburgazzoli** [comment]
> We don't need this since the namespace is not passed/used in the model registry controller,

Even better then
 
> We can have defaulting webhook set the default values But then we can't have the default value in the CRD description if it's not in the yaml.

I know, but it seems better than having diverging CRDs (IMHO).  I'm not very sure to see the value of having such default made visible since it is optional, it may be better to report it in the status, but this is another discussion and as said, I'm fine with this implementation.

File: config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (VaishnaviHire) [inline config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml:454]: Should this be a required field? Any addition of required field will need DSC api version update.

Can we follow same cadence as other fields , where the fields are optional with default values set

Comment (dhirajsb) [reply config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml:454]: The primary requirement is that Dashboard MUST have a value in this property so they know which model registry namespace to use. So, it can't be an optional/empty value. I looked, but we don't seem to have a Default() implementation to provide 
If the new required property has a default value, k8s seems to add it automatically when the CRD is updated. So, is an api version update still needed?

Comment (dhirajsb) [reply config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml:454]: I did find the common webhook in `admission/webhook.go`, but we aren't using it to set any default values it seems.

**Comment by dhirajsb** [comment]
@VaishnaviHire @zdtsw I'm seeing some strange reconcile errors in `generateCertSecret()` when I tried running e2e tests in my local cluster. I've captured them in a [gist](https://gist.github.com/dhirajsb/7b6d111d88f7097c6e510a9c868d38a3). 

It appears the controller is just not seeing an existing model registry secret `default-modelregistry-cert` in the `istio-system` namespace and tries to create it and fails reconcile over and over. 

I've verified that the operator serviceaccount can actually get that secret:
```shell
[dbokde@dbokde model-registry-operator (feat/auth-config-env =)]$ kc get secrets -n istio-system default-modelregistry-cert --as system:serviceaccount:opendatahub-operator-system:opendatahub-operator-controller-manager
NAME                         TYPE     DATA   AGE
default-modelregistry-cert   Opaque   2      14d
```

Have you seen this kind of error before? I know the client caches some k8s resources, but this same issue doesn't happen with `knative-serving-cert` secret for some reason. :open_mouth: 

I wonder if the error @tonyxrmdavidson saw in QE cluster was something similar??

**Comment by dhirajsb** [comment]
I found the issue, there is a missing cache config for `default-modelregistry-cert` just like the one for [knative-serving-cert](https://github.com/opendatahub-io/opendatahub-operator/blob/5759f5e28a8e5836b93966a6bd1a7377bf270530/main.go#L324).

**Comment by dhirajsb** [comment]
@zdtsw saw that you fixed the secrets watch issue already in https://github.com/opendatahub-io/opendatahub-operator/pull/1223 :smiley:

**Comment by zdtsw** [comment]
> @zdtsw saw that you fixed the secrets watch issue already in #1223 😃

yeah, i saw this problem 2 days ago when i was changing e2e test, but did not think it was caused by the code.
till yesterday got a report on it. 
openshift-ci was not statble (as usual) thats why it took hours to finally get it merged, sorry for the problem.

**Comment by zdtsw** [comment]
> @VaishnaviHire @zdtsw I'm seeing some strange reconcile errors in `generateCertSecret()` when I tried running e2e tests in my local cluster. I've captured them in a [gist](https://gist.github.com/dhirajsb/7b6d111d88f7097c6e510a9c868d38a3).
> 
> It appears the controller is just not seeing an existing model registry secret `default-modelregistry-cert` in the `istio-system` namespace and tries to create it and fails reconcile over and over.
> 
> I've verified that the operator serviceaccount can actually get that secret:
> 
> ```shell
> [dbokde@dbokde model-registry-operator (feat/auth-config-env =)]$ kc get secrets -n istio-system default-modelregistry-cert --as system:serviceaccount:opendatahub-operator-system:opendatahub-operator-controller-manager
> NAME                         TYPE     DATA   AGE
> default-modelregistry-cert   Opaque   2      14d
> ```
> 
> Have you seen this kind of error before? I know the client caches some k8s resources, but this same issue doesn't happen with `knative-serving-cert` secret for some reason. 😮
> 
> I wonder if the error @tonyxrmdavidson saw in QE cluster was something similar??

 the caching part was started months way back when modelreg did not require this secret yet, but the PR got splitted and rebases many times even after recently changes from modelreg. 
so the cache secret part did not get updated accordingly with modelreg secret. 
but now i start to think, could there be some hidden problem in our e2e.... esp the ones got merged/passed e2e between the "caching" till the "secret fix": modelreg should be kept reconciled, how DSC status end with Ready 🤔 
i will need dig into this more today

**Comment by lburgazzoli** [comment]
LGTM

**Comment by dhirajsb** [comment]
@zdtsw thanks for being on top of this. I did learn a lot more about the odh operator going through the troubleshooting in my own cluster. 
I'm investigating what's going on with the failing e2e tests in CI. I tried it a few times locally, and e2e tests are passing in my personal cluster.

**Comment by dhirajsb** [comment]
@zdtsw looking at the raw e2e log I see the warning:
```shell
=== RUN   TestOdhOperator/create_Opendatahub_components/e2e-test-dsc/Creation_of_DataScienceCluster_instance
2024-09-06T05:07:20Z	INFO	KubeAPIWarningLogger	unknown field "spec.components.modelregistry.registriesNamespace"
```
Along with several others about other components. 
Is the DSC manifest not being updated from the start in the e2e test?

**Comment by dhirajsb** [comment]
Forcing re-test all to have it re-build image with the secret cache fix.

**Comment by dhirajsb** [comment]
@zdtsw @VaishnaviHire I am not sure why, but the same e2e test is able to set the default value in my local cluster and not in the openshift ci. I think we also need the validating webhook to set the default value, in case the value is an empty string. I was hoping using `scheme.Default()` would do it, but it looks like it's only doing yaml schema validation and not higher level CRD default values. 
Having said that, an existing DSC in my cluster without the new property automatically got the default value set correctly. So, it looks like k8s API controller behaves differently than the generated client API code. 

I'll work on those fixes and additional checks for the missing default value.

**Comment by dhirajsb** [comment]
I found an [issue](https://github.com/kubernetes-sigs/controller-tools/issues/893) in kubebuilder apiserver that basically says defaultvalues have to be omitempty fields so it can detect that the value is missing. Which makes it always have a value at runtime, so _required_ in effect! :shrug:

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [inline components/modelregistry/modelregistry.go:51]: why we set omitempty here?
why not `RegistriesNamespace string json:"registriesNamespace"``

then when we create https://github.com/opendatahub-io/opendatahub-operator/blob/602b5bb49bd018ddff126ac9138db0856b73efb4/tests/e2e/helper_test.go#L170 without set any value, it will get the one from default as `odh-model-registries `

Comment (zdtsw) [reply components/modelregistry/modelregistry.go]: with https://github.com/dhirajsb/opendatahub-operator/commit/b2b08b582ae8a24ae660c427a6696b28e53648e3 change. 
test with this DSC to simulate our e2e test 
```
kind: DataScienceCluster
apiVersion: datasciencecluster.opendatahub.io/v1
metadata:
  labels:
    app.kubernetes.io/created-by: opendatahub-operator
    app.kubernetes.io/instance: default
    app.kubernetes.io/managed-by: kustomize
    app.kubernetes.io/name: datasciencecluster
    app.kubernetes.io/part-of: opendatahub-operator
  name: default-dsc
spec:
  components:
    modelregistry:
      managementState: Managed
```
after DSC is created,
```
spec:
  components:
    codeflare: {}
    kserve:
      serving:
        ingressGateway:
          certificate:
            type: OpenshiftDefaultIngress
        managementState: Managed
        name: knative-serving
    modelregistry:
      managementState: Managed
      registriesNamespace: odh-model-registries
    trustyai: {}
    ray: {}
    kueue: {}
    workbenches: {}
    dashboard: {}
    modelmeshserving: {}
    datasciencepipelines: {}
    trainingoperator: {}
```

**Comment by zdtsw** [comment]
IT PASSED FINALLY :cry:

**Comment by dhirajsb** [comment]
:rofl: I am not exactly sure what changed from when it was running e2e tests without `omitempty` before, but I'm happy it finally passed! :champagne:

File: bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (VaishnaviHire) [inline bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:455]: Hi @etirelli @lburgazzoli @zdtsw this probably is a non issue, but I would like to confirm if adding required fields in DSC is a compatible change  ? Ref: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility .

Comment (zdtsw) [reply bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:455]: your concern is on 
1. we should uplift API version ?
2. if old DSC CR (esp for modelreg as component) is compatible with this new changes?

Comment (VaishnaviHire) [reply bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:455]: Right both.

Also for 1. If we should update API version or remove this from required fields

**Comment by zdtsw** [comment]
btw @dhirajsb , one more thing would like to confirm:
when we say "This property should be configurable by the user at DSC creation time," 
do you mean: 
- for a brand new installation of RHOAI,?
or do you mean:
- when the time ModelReg as a component is enabled in DSC? 

Just want to be crystal clear , with this PR,  Operator works with
- RHOAI new installation
- RHOAI upgrade 
these two are handling component config differently.

Any user if already have RHOAI running in their cluster, after upgrade Operator, they would have ModelReg by default as "removed" but the value `registriesNamespace: odh-model-registries` is already set in DSC CR once upgrade is done.
There is no  way to change registriesNamespace afterwards, Unless they delete existing DSC CR and create a brand new DSC CR with any namespace for registriesNamespace

Is this what you would like?

File: config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (VaishnaviHire) [reply config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml:454]: Getting more inputs here https://github.com/opendatahub-io/opendatahub-operator/pull/1221#discussion_r1750485311

**Comment by dhirajsb** [comment]
> when we say "This property should be configurable by the user at DSC creation time,"
do you mean:
for a brand new installation of RHOAI,?
or do you mean:
when the time ModelReg as a component is enabled in DSC?

That's a good question. The way it works right now, it actually gets set when the DSC itself is created, not when the model registry component is enabled. Honestly, I'm not sure it's a business requirement to have users be able to configure it when enabling model registry right now. 
If users ask for such a dynamic behavior, there are ways to create an appropriate validating hook, but we can defer that for a future version.

File: bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (dhirajsb) [reply bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:455]: Doesn't marking it optional with `omitempty`, but with a default value make the API server always set a value?

**Comment by dhirajsb** [comment]
I am testing the more complex use case with cel validation to allow users to set the namespace when model registries was already enabled with: 
```go
// ModelRegistry struct holds the configuration for the ModelRegistry component.
// The property `registriesNamespace` is immutable when management state was not equal to `Removed`
// +kubebuilder:object:generate=true
// +kubebuilder:validation:XValidation:rule="(self.managementState == 'Removed' || oldSelf.managementState == 'Removed') || (self.managementState != 'Removed' && self.registriesNamespace == oldSelf.registriesNamespace)",message="RegistriesNamespace is immutable when ManagementState was not equal to Removed"
type ModelRegistry struct {
	components.Component `json:""`

	// Namespace for model registries to be installed, configurable once when model registry is enabled, defaults to "odh-model-registries"
	// +kubebuilder:validation:Required
	// +kubebuilder:default="odh-model-registries"
	RegistriesNamespace string `json:"registriesNamespace,omitempty"`
}
```
I'll push this change if it works in my cluster.

**Comment by dhirajsb** [comment]
Pushing the new cel validation rule to test it in CI in parallel since it takes such a long time.

File: components/modelregistry/modelregistry.go
Comment (lburgazzoli) [inline components/modelregistry/modelregistry.go:198]: IMHO, a controller should never touch the spec, usually those defaults should end up being rendered in the status but I let @zdtsw and @VaishnaviHire to review.

Comment (zdtsw) [reply components/modelregistry/modelregistry.go:198]: we do have some patching but not on DSCI/DSC CR only for the component(i.e kserve) related resources thro templating.
this patching to DSC, would not it cause extra unnecessary reconcile from DSC controller?

File: bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (dhirajsb) [reply bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:455]: Based on my testing, K8s considers an empty string value "" valid for this optional field and doesn't use the default value. That's why @zdtsw had to add it to the test DSC create func. 
But, @VaishnaviHire has a point about needing to change API versions if we add a new required property. So, we should avoid that. That's why I'm trying to get it to always have the property, without explicitly setting it to required. There are other properties like KNative Serving Name that also work in a similar fashion.

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [reply components/modelregistry/modelregistry.go]: unless i am wrong for the knative serving default name "knative-serving", does not matter kserve is enabled or removed, that value knative-serving is always set in the kserve.serving.name and no limitation saying user cannot change its value to something else e.g knative-serving2

Comment (zdtsw) [reply components/modelregistry/modelregistry.go]: case 1
```
 kserve:
      managementState: Managed
      serving:
        ingressGateway:
          certificate:
            type: OpenshiftDefaultIngress
        managementState: Managed
        name: ''
```
this ^ can be fine to create DSC CR, but will be problem for kserve to use knative due to some hardcoded logic



case 2
```
    kserve:
      managementState: Managed
      serving:
        ingressGateway:
          certificate:
            type: OpenshiftDefaultIngress
        managementState: Managed
        name:
```
this ^ will generate a DSC CR as:
```
kserve:
      managementState: Managed
      serving:
        ingressGateway:
          certificate:
            type: OpenshiftDefaultIngress
        managementState: Managed
        name:knative-serving
```

case 3
and with this config
```
    kserve:
      managementState: Managed
      serving:
        ingressGateway:
          certificate:
            type: OpenshiftDefaultIngress
        managementState: Managed
```
DSC CR gets
  ```
  kserve:
      managementState: Managed
      serving:
        ingressGateway:
          certificate:
            type: OpenshiftDefaultIngress
        managementState: Managed
        name: knative-serving
```

Comment (dhirajsb) [reply components/modelregistry/modelregistry.go]: I think case2 was the problem with modelregistries.registriesNamespace. I.e. if you create a DSC with:
```yaml
modelregistry:
  registriesNamespace: ""
```
It is considering a valid value, and obviously causes errors. 
Hence, the need for a defaulting webhook to catch this case and patch it to the default value.

File: controllers/datasciencecluster/datasciencecluster_controller.go
Comment (lburgazzoli) [inline controllers/datasciencecluster/datasciencecluster_controller.go:321]: this can probably be moved in the `UpdateWithRetry` block below so to update it only when the component reconciliation succeeded.

File: main.go
Comment (lburgazzoli) [inline main.go:353]: @zdtsw  I guess we can use the `modelregistry.DefaultModelRegistriesNamespace` variable as it is going to be amended for downstream ?

Comment (zdtsw) [reply main.go:353]: yep, i have some tests ongoing and will have this updated.

File: controllers/datasciencecluster/datasciencecluster_controller.go
Comment (zdtsw) [reply controllers/datasciencecluster/datasciencecluster_controller.go:321]: will get into with changes in cache name as well

File: main.go
Comment (zdtsw) [reply main.go]: updated

File: controllers/datasciencecluster/datasciencecluster_controller.go
Comment (zdtsw) [reply controllers/datasciencecluster/datasciencecluster_controller.go]: updated

**Comment by lburgazzoli** [comment]
@zdtsw I guess we should squash the commits before merge right ?

**Comment by zdtsw** [comment]
> @zdtsw I guess we should squash the commits before merge right ?

yes,  it will be squahsed before the final one commits to not "pollute" git log :sweat:

**Comment by ykaliuta** [comment]
> @zdtsw I guess we should squash the commits before merge right ?

It is actually merging policy preventing making proper patchsets.

**Comment by VaishnaviHire** [comment]
/lgtm

Tested Upgrade from ODH 2.17 --> latest

**Review by zdtsw — APPROVED** [review]

File: controllers/webhook/webhook.go
Comment (zdtsw) [reply controllers/webhook/webhook.go:143]: i think the main reason to use 2 differnet endpoints , one per each, instead of having one endpoint for both (which is applicable as well)  is we already had validation webhook in downstream. By this way, we can get the downstream part out first, and the mutating part as a separate one to test. 
another reason is, it is slightly easy to maintain, the logic will be simple in its own Handle() and mutating only need on DSC not DSCI  for now.

Comment (ykaliuta) [reply controllers/webhook/webhook.go:143]: > According to the k8s documentation:
> 
> ```
> When allowing a request, a mutating admission webhook may optionally modify the incoming object as well. 
> ```
> 
> So, we need a mutating webhook to modify a CR. They also use different sets of CRs and verbs.

Do I understand right that you claim that it is impossible?

Comment (dhirajsb) [reply controllers/webhook/webhook.go:143]: How are you reading the documentation? Please point me to the documentation where it says a validating webhook can respond with a patch, maybe I missed it.

Comment (bartoszmajsak) [reply controllers/webhook/webhook.go:143]: Have you considered using [`admission.CustomDefaulter`](https://github.com/kubernetes-sigs/controller-runtime/blob/2eb879f25c4829825e6f2511cda035de41ec7030/pkg/webhook/admission/defaulter_custom.go) instead? It seems like a good fit if we want to follow the single responsibility principle. It would also simplify the code.

Comment (ykaliuta) [reply controllers/webhook/webhook.go:143]: > How are you reading the documentation? Please point me to the documentation where it says a validating webhook can respond with a patch, maybe I missed it.

I'm just asking questions :)

Comment (ykaliuta) [reply controllers/webhook/webhook.go:143]: > i think the main reason to use 2 differnet endpoints , one per each, instead of having one endpoint for both (which is applicable as well) is we already had validation webhook in downstream. By this way, we can get the downstream part out first, and the mutating part as a separate one to test. another reason is, it is slightly easy to maintain, the logic will be simple in its own Handle() and mutating only need on DSC not DSCI for now.

I did some experiments and agree, it makes logic more simple.
But I have another question. It has a check for dsc.Spec.Components.ModelRegistry.RegistriesNamespace but with // +kubebuilder:default="odh-model-registries" it never true for me.

Comment (dhirajsb) [reply controllers/webhook/webhook.go:143]: I considered `CustomDefaulter`, but it doesn't pass along the previous version of the object. It's nice to have it if the defaulting logic ever becomes complex enough to know what the old spec was. 
We are already using complex validation logic in the MR CEL validator. So, I don't know whether defaulting webhook will ever need to also do that, but I also don't know that it never will. :smile:

**Pull request merged**
